cpp/src/barretenberg/honk/pcs/gemini/gemini.cpp

```diff
@@ -0,0 +1,365 @@
+ * @param batched_unshifted F(X) = ∑ⱼ ρʲ   fⱼ(X)
+ * @param batched_to_be_shifted G(X) = ∑ⱼ ρᵏ⁺ʲ gⱼ(X)
+ * @return std::vector<Polynomial>
+ */
+template <typename Params>
+std::vector<typename barretenberg::Polynomial<typename Params::Fr>> MultilinearReductionScheme<
+    Params>::compute_fold_polynomials(std::span<const Fr> mle_opening_point,
+                                      Polynomial&& batched_unshifted,
+                                      Polynomial&& batched_to_be_shifted)
+{
+
+    using Fr = typename Params::Fr;
+    using Polynomial = barretenberg::Polynomial<Fr>;
+
+    const size_t num_variables = mle_opening_point.size(); // m
+
+    const size_t num_threads = get_num_cpus_pow2();
+    constexpr size_t efficient_operations_per_thread = 64; // A guess of the number of operation for which there
+                                                           // would be a point in sending them to a separate thread
+
+    // Allocate space for m+1 Fold polynomials
+    //
+    // The first two are populated here with the batched unshifted and to-be-shifted polynomial respectively.
+    // They will eventually contain the full batched polynomial A₀ partially evaluated at the challenges r,-r.
+    // This function populates the other m-1 polynomials with the foldings of A₀.
+    std::vector<Polynomial> fold_polynomials;
+    fold_polynomials.reserve(num_variables + 1);
+
+    // F(X) = ∑ⱼ ρʲ fⱼ(X) and G(X) = ∑ⱼ ρᵏ⁺ʲ gⱼ(X)
+    Polynomial& batched_F = fold_polynomials.emplace_back(std::move(batched_unshifted));
+    Polynomial& batched_G = fold_polynomials.emplace_back(std::move(batched_to_be_shifted));
+    constexpr size_t offset_to_folded = 2; // Offset because of F an G
+    // A₀(X) = F(X) + G↺(X) = F(X) + G(X)/X.
+    Polynomial A_0(batched_F);
+    A_0 += batched_G.shifted();
+
+    // Allocate everything before parallel computation
+    for (size_t l = 0; l < num_variables - 1; ++l) {
+        // size of the previous polynomial/2
+        const size_t n_l = 1 << (num_variables - l - 1);
+
+        // A_l_fold = Aₗ₊₁(X) = (1-uₗ)⋅even(Aₗ)(X) + uₗ⋅odd(Aₗ)(X)
+        fold_polynomials.emplace_back(Polynomial(n_l));
+    }
+
+    // A_l = Aₗ(X) is the polynomial being folded
+    // in the first iteration, we take the batched polynomial
+    // in the next iteration, it is the previously folded one
+    auto A_l = A_0.data();
+    for (size_t l = 0; l < num_variables - 1; ++l) {
+        // size of the previous polynomial/2
+        const size_t n_l = 1 << (num_variables - l - 1);
+
+        // Use as many threads as it is useful so that 1 thread doesn't process 1 element, but make sure that there is
+        // at least 1
+        size_t num_used_threads = std::min(n_l / efficient_operations_per_thread, num_threads);
+        num_used_threads = num_used_threads ? num_used_threads : 1;
+        size_t chunk_size = n_l / num_used_threads;
+        size_t last_chunk_size = (n_l % chunk_size) ? (n_l % num_used_threads) : chunk_size;
```

Comment:
I think this robustness is nice to have but I wonder if in practice we can always assume (otherwise, enforce) that num_threads is a power of 2 and thus divides `n_l`.

Reply:
Lots of 6-core laptops out there)